### PR TITLE
Add track templates

### PR DESCRIPTION
**Pull request by veidenberg**
### Description
Add track payload templates and corresponding script to load tracks for the beta-2 update and beyond.

It has 2 parts:
1) Track templates: each track type has it's own set of values, some of which are internal (e.g. genome browser settings, known by webteam) some external (e.g. track description or datafile name, known by datateams/production). The track templates in this repository is a central place where these fields can be updated and then used to populate Track API.
2) Track loading script: whenever new track datafiles are generated/updated and copied to genome browser NFS area, production can use this script to register corresponding tracks in Track API (to make the tracks available for Ensembl client). The script will read the input datafiles directory and use the corresponding track payload templates to submit the tracks to Track API. The script can be called on specific species/tracks or all available datafiles (see the command-line help text). Web will use this script to load all the tracks for beta-2 update (since the datafiles have already been generated).

### Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-2673

### To test

```bash
[ens_adm02@codon-dm-01 ~]$ cd trackapi/utils; export TRACK_API_URL=https://dev-2020.ensembl.org/api/tracks
[ens_adm02@codon-dm-01 utils]$ export TRACK_DATA_DIR=<nfs_path_to_GB_files> # optional in this case
[ens_adm02@codon-dm-01 utils]$ ./submit_track_templates.py -g a7335667-93e7-11ec-a39d-005056b38ce3 -t gc -o
```
This will replace the gc track for pig in dev (you can confirm by comparing the track ID in https://dev-2020.ensembl.org/api/tracks/track_categories/a7335667-93e7-11ec-a39d-005056b38ce3 before and after the script run). Add `-d` to skip track submission and just print the track payload.

### CODE REVIEW

File: templates/transcripts-gene-pc-rev.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+additional_info: Reverse strand
+colour: DARK_GREY
+display_order: 100
+on_by_default: true
+description: Shows all protein coding genes on the reverse strand. <annotation_description>
+sources:
+  - name: <annotation_method>
+    url: <annotation_url>
+datafiles:
+  gene: transcripts.bb
```
Comment (azangru) [inline templates/transcripts-gene-pc-rev.yaml:20]: github complains about this file; possibly because the next line is both empty and indented:

image

this is different from all other yaml templates; probably makes sense to clean this up

File: templates/gerp-elements.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  track_category_id: assembly
+  label: Assembly
+  type: Genomic
+trigger:
+  - track
+  - expand
+type: regular
+label: Constrained elements
+display_order: 960
+on_by_default: true
```
Comment (azangru) [inline templates/gerp-elements.yaml:12]: Was there any decision about whether all of these tracks should be on by default?

_(I don't remember whether the client respects this flag; but that's another matter.)_

File: templates/regulation.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+trigger:
+  - track
+  - regulation
+type: regular
+label: Regulatory annotation
+display_order: 890
+on_by_default: true
+description: |-
+  Promoters, enhancers and open chromatin regions are available for all species. 
+  Transcription factor binding and CTCF binding sites are only available for human and mouse.
```
Comment (azangru) [inline templates/regulation.yaml:15]: Do you know why we keep writing this in the description? I don't think we are even showing this anywhere in the genome browser... Also, I can't remember whether we hide the "TF binding" and the CTCF colour blocks on genomes other than human and mouse. Probably not?

image

File: templates/transcripts-gene-other-rev.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  - gene-other-rev
+type: gene
+label: Other genes
+additional_info: Reverse strand
+colour: GREY
+display_order: 110
+on_by_default: true
+description: Shows all non-coding genes on the reverse strand. <annotation_description>
+sources:
+  - name: <annotation_method>
```
Comment (azangru) [inline templates/transcripts-gene-other-rev.yaml:17]: How do these get replaced?

File: utils/submit_tracks_beta2.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      UUID(subdir, version=4)
+    except ValueError:
+      print(f"Skipping non-genome directory {subdir}")
+      continue
+    if args.genome and subdir not in args.genome:
+      continue
+    print(f"Processing genome {subdir}")
+    for file in os.listdir(f"{root_dir}/{subdir}"):
+      if file.endswith('.bb') or file.endswith('.bw'):
+        find_template(subdir, file)
```
Comment (azangru) [inline utils/submit_tracks_beta2.py:44]: Naming. The name `find_template` suggests that it is going to be a function that returns the found template. But it doesn't. It immediately applies the template, and returns nothing.

File: utils/submit_tracks_beta2.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+# 2) Load the corresponding track payload template
+def find_template(genome_id: str, datafile: str) -> None:
+  if args.file and datafile not in args.file:
+    return
+  filename = os.path.splitext(datafile)[0]
+  template_files = [os.path.basename(f) for f in glob.glob(f"{template_dir}/*.yaml")]
+  if filename in template_files: # 1-to-1 match
+    return apply_template(genome_id, datafile, filename)
```
Comment (azangru) [inline utils/submit_tracks_beta2.py:53]: This is a purely stylistic comment from someone who isn't familiar with the conventions of the language. `apply_template` doesn't return anything. When the code reads `return apply_template(genome_id, datafile, filename)`, doesn't it make you think that `apply_template` returns something useful? If instead you wrote 

```
apply_template(genome_id, datafile, filename)
return
```

wouldn't the intent be clearer?

File: utils/submit_tracks_beta2.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  submit_track(track_data)
+
+# 4) Submit the track payload to Track API
+def submit_track(track_data: dict, second_try: bool = False) -> None:
+  print(f"Submitting track: {track_data['label']}")
+  if args.dry_run:
+    print(track_data)
+    return
+
+  track_api_url = f"https://{args.env}-2020.ensembl.org/api/tracks"
```
Comment (azangru) [inline utils/submit_tracks_beta2.py:75]: Shouldn't  the host (`https://{args.env}-2020.ensembl.org`) come from the environment variable as a whole. Doesn't the hardcoded `-2020` look dangerous?

File: templates/gerp-elements.yaml (hunk 2 above)
Comment (veidenberg) [reply templates/gerp-elements.yaml:12]: This is why we have templates here: we (or data teams or product owner etc) can review and edit properties like `on_by_default` and it will be used the next time data loading script is run. Whether the constrained elements track should to be on by default, I don't know. Who do we ask?

File: templates/regulation.yaml (hunk 3 above)
Comment (veidenberg) [reply templates/regulation.yaml:15]: The regulation track description field comes from [here](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1969). But we also have feature-specific descriptions hardcoded in the client, don't we?

File: templates/variant-ensembl-details.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      - label-snv-id
+      - label-snv-alleles
+      - label-other-id
+      - label-other-alleles
+      - show-extents
+type: variant
+label: Ensembl short variants
+display_order: 1001
+on_by_default: true
+description: All short variants (SNPs and indel) data from Ensembl - e110
```
[the remark this answers is not in the document]
Comment (veidenberg) [reply templates/variant-ensembl-details.yaml:21]: That's what I currently have from Variation (and variation tracks don't get updates for beta2). When the description becomes outdated, it's meant to be Variation's task to update the template (before they hand over datafiles which triggers the Track API loading).

File: utils/submit_tracks_beta2.py (hunk 5 above)
Comment (veidenberg) [reply utils/submit_tracks_beta2.py:44]: Suggestions? I divided the script to steps (so that production can switch any part, e.g. feed the list of input datafiles from another source than a data dir). This step matches a datafile to its track template and continues with the next step.

Comment (azangru) [reply utils/submit_tracks_beta2.py:44]: > Suggestions?

Something like:
```
      if file.endswith('.bb') or file.endswith('.bw'):
        create_track_for_file(subdir, file)
```
?

File: templates/repeats.dust.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+label: "Low complexity: Dust"
+display_order: 984
+on_by_default: true
+description: Shows low complexity regions (regions of the genome with a biased distribution of nucleotides, such as a repeat) on the genome, identified by Dust
+sources:
+  - name: Dust
+    url: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-015-0654-5
+datafiles:
+  repeat-details: repeats.dust.bb
+  repeat-summary: ""
```
[the remark this answers is not in the document]
Comment (veidenberg) [reply templates/repeats.dust.yaml:24]: It tells the GB to switch to another eard (summary view) on zoom out levels (see the settings) but we don't have a datafile for that atm. Hence the empty track for repeats when zoomed out.

[the remark this answers is not in the document]
Comment (veidenberg) [reply templates/repeats.dust.yaml:24]: Track loading script only cares about the `details` file (see the name of the template file) since for each zoomin/zoomout pair the track only needs to be registered once.

File: utils/submit_tracks_beta2.py (hunk 5 above)
Comment (veidenberg) [reply utils/submit_tracks_beta2.py:44]: So I should merge `find_template` and `apply_template` (and `submit_track`) to `create_track_for_file` function?

Comment (azangru) [reply utils/submit_tracks_beta2.py:51]: But you are already calling `submit_track` from `apply_template`, which in turn you call from within `find_template`. 🤷‍♂️ You could continue doing the same, but rename the top function to something like `create_track_for_file `.

An alternative suggestion is:

```
if file.endswith('.bb') or file.endswith('.bw'):
  template = find_template(subdir, file)
  if template:
    genome_id = subdir
    apply_template(genome_id, file, template)
```

**Comment by jyothishnt** [comment]
I am not good in python so I will leave others to comment on the code, but I had gone through it for my understanding. I dont know how to test it though. because the files are not accessible from my laptop.

**Comment by veidenberg** [comment]
> Hi @veidenberg in `datafiles` section I see 3 different entries. One just details with .bb file, one that contain both .bw and bb. Na done that contain bb and summary="". I read your comment about using 2 files for different zoom levels. But shouldn't it be consistent across?

The keys/values in `datafiles` define the GB eard programs and datafiles used to draw the track. The different template configurations reflect the corresponding implementations in GB:
1) details + summary + both bb/bw files defined: the track switches the eard program and the datafiles for different zoom levels (e.g. variation track).
2) details + summary + only one datafile defined: the track switches the eard program for different zoom levels, but one of these eards has no datafile = empty track for that zoom level (e.g. repeats track)
3) only "details" (or "summary", or any other key value) defined: no switchover, single eard is used at all zoom levels. Some tracks (e.g. compara) happens to have this single program named "summary" or "details" because of historic reasons (these eard programs display the summary/details view, but we decided to split it to separate tracks and show both views at all zoom levels).
In summary, the `datafiles` keys/values are consistent in schema (implementation) point of view, but renaming the compara track eard (and template key) names probably makes sense to avoid confusion.

**Comment by veidenberg** [comment]
> I am not good in python so I will leave others to comment on the code, but I had gone through it for my understanding. I dont know how to test it though. because the files are not accessible from my laptop.

I added a testing example to the PR description.

File: utils/submit_tracks_beta2.py (hunk 5 above)
Comment (veidenberg) [reply utils/submit_tracks_beta2.py]: Some datafiles will apply/submit multiple templates/tracks, hence the dedicated function for file=>template matching. I renamed it to `match_template` if it's less confusing (although `find_template` is anyway less clear about returning anything than say, `get_template`).

**Review by kamaldodiya — APPROVED** [review]

File: templates/regulation.yaml (hunk 3 above)
Comment (azangru) [reply templates/regulation.yaml:15]: > Where do these feature-specific descriptions come from, is it hardcoded in the client?

Yes, they are hard-coded on the client. These feature-specific descriptions look visually the same as tracks, but aren't tracks, and thus can't be using the track api.

https://github.com/Ensembl/ensembl-client/blob/80360adb0a6117f77753ebb0e2a5457c7ec29b36/src/content/app/genome-browser/constants/regulationLegend.ts#L51

I also found where we are showing the description of the regulation track:

image

It is still a bit odd that we write about "human and mouse" regardless of the genome that we are showing 🤷‍♂️

**Review by jyo-nt — APPROVED** [review]

**Comment by veidenberg** [comment]
Merged to dev in order to test the track settings field in dev db. Feel free to keep adding comments on templates and loader script.
